### ISSUE TITLE
Type guards with falsy NamedTuples can narrow a NamedTuple to Never

### ISSUE THREAD

**Issue by markjoshwel** [microsoft/pyright#11019]
**Describe the bug**

When in a negative conditional branch, NamedTuple classes with a custom `__bool__` narrows the type to `Never`, suggesting unreachable code when it is actually reachable.

This issue seems to occur on my NamedTuple-inheriting classes, whereas plain classes narrow correctly.

**Code or Screenshots**

```python
from typing import reveal_type, NamedTuple


class Example():
    field: str = "a"

    def __bool__(self) -> bool:
        return False


class ExampleButNamedTuple(NamedTuple):
    field: str = "b"

    def __bool__(self) -> bool:
        return False


ex = Example()
if ex:
    reveal_type(ex.field)  # Type of "ex.field" is "str"
if not ex:
    reveal_type(ex.field)  # Type of "ex.field" is "str"

exnt = ExampleButNamedTuple()
if exnt:
    reveal_type(exnt.field)  # Type of "ex.field" is "str"
if not exnt:
    reveal_type(exnt.field)  # Type of "ex.field" is "Never" (should be "str")

```

**VS Code extension or command-line**

Running Pyright via [Pyright Playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQDRQByZEhAJgCoCuCFhAKCEBjCmQDO4qAFEAHqz6EAFAEoAXAKhaowJIQrs1UcTBBQAvFABEZK0O1R2hYFCpUARmDDUqS8fuAVKABaAD4oT28NBwciGG4QFCgAMUp-IQFRCSk5BX4AIW4YFjYuXn4lEo4eRXVNbV19Q2NTC2t3O3qtJxc3SJ8-AKCwiK8KaJitOISk1Ip0oUJZNtzsCpUBJBcliamScmpaZSWAOkaDIKgAYihOREIoMBcrU-P2K0wpKxMQOy2oFBgfA7LoEfaUGj3JSvPQXLQ3O50R7PGFND5IL4-TpLFD4SyrRSFYqsarlZQbf44mC7MGkCFHaGyXFnWHsS4I%2B7I6yogzozGmP4uQHApnU0FEOmHKFUllNdm3TlPbmyWW8z7WJj7X4CIA)
- Pyright Version 1.1.406
- Python Version 3.13

**Comment by markjoshwel** [comment]
I've attempted to fix this, as per the commit above, and apologies as this is my first time trying out TypeScript, but if my issue is not a duplicate and my commit is in the right direction, I'd happily open a PR. The commit fixes my use case and passes all tests. Any input is appreciated, thanks!

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

You're changing the interface contract for a `tuple` here in your override of `__bool__`. I think it's reasonable for a type checker to assume that semantics for built-in classes are maintained when overridden.
